### PR TITLE
[WIP] Proposal: Refactor Registration Scripts

### LINKED ISSUE #109: RegisterPluginAndStartStreamDeck.* Scripts Get Out of Sync Easily
The RegisterPluginAndStartStreamDeck.* scripts are duplicated in multiple spots. This causes them to get out of sync easily as they are updated.

### PR DESCRIPTION
**Pull request by NotMyself**
Fixes #109 

Over the last couple of weeks, we have seen the `RegisterPluginAndStartStreamDeck.ps1` and `RegisterPluginAndStartStreamDeck.sh` files get out of sync or have changes get reverted. This is caused by the scripts being duplicated in both the `SamplePlugin` and `Templates/StreamDeck.PluginsTemplate.Csharp` directories.

This proposal centralizes the placement of all scripts into a `scripts` directory located at the root of the project. For this to work, the following requirements must be met:

- [x] No duplication of script files.
- [x] `SamplePlugin` should build from the command line or Visual Studio from a clean clone of the repository.
- [x] `StreamDeck.PluginTemplate.Csharp` nuget package should contain the registration scripts when built.
- [x] Developers should be able to install the template locally for testing
- [x] Ensure the registration scripts located in the template directory are ignored by git.

To meet these requirements, the following changes are proposed:

1. All scripts are relocated to a single `scripts` directory.
2. `SamplePlugin` has been modified to include the scripts from the new location.
3. `StreamDeckPluginTemplate.nuspec` has been updated to include the scripts on a build.
4. Both an `install.sh|ps1` and `uninstall.sh|ps1` scripts are added to the scripts directory to facilitate local development.

I have tested all of these requirements on OSX in both `bash` and `powershell core`. But it would be a good idea to have a windows user perform the same tests.

Tests:

1. Run a `dotnet build` from the `SamplePlugin` directory.
   - Expected Result: build should complete, plugin should be copied to plugin directory and Stream Deck app should be restarted.
2. Run `nuget pack Templates/StreamDeck.PluginTemplate.Csharp/StreamDeckPluginTemplate.nuspec` from the root of the repository.
   - Expected Result: the registration scripts should be included in the build nuget package.
3. Install the template locally by running `./scripts/install.sh|ps1` from the root of the repository.
   - Expected Result: The template should install correctly and new projects created with it should execute the scripts as expected.
4. Uninstall the template locally by running `./scripts/uninstall.sh|ps1`.
   - Expected Result: The template should be cleanly removed.

**Note:** For test 2, you may need to temporarily edit the nuspec file to have a valid version number.

### CODE REVIEW

File: scripts/registration/RegisterPluginAndStartStreamDeck.ps1 [hunk 1]
```diff
@@ -4,13 +4,13 @@ Write-Host "Script root: $PSScriptRoot`n"
...
 
-if (!($PSSCriptRoot)) {
+if ($PSSCriptRoot.Length -eq 0) {
   $basePath = $PWD.Path;
 }
 
 
 # Load and parse the plugin project file
-$pluginProjectFile = Join-Path $basePath "SamplePlugin.csproj"
+$pluginProjectFile = "$basePath\_StreamDeckPlugin_.csproj"
```
Comment (NotMyself) [inline scripts/registration/RegisterPluginAndStartStreamDeck.ps1]: This is going to be a problem. Is there a better way to do this that will work in both contexts?

File: scripts/registration/RegisterPluginAndStartStreamDeck.ps1 [hunk 2]
```diff
@@ -20,46 +20,42 @@ $buildConfiguration = "Debug"
...
-}
+$streamDeckExePath = "$($ENV:ProgramFiles)\Elgato\StreamDeck\StreamDeck.exe"
+
+# For now, this PS script will only be run on Windows.
+$bindir = "$basePath\bin\Debug\$targetFrameworkName\win-x64"
 
 # Make sure we actually have a directory/build to deploy
 If (-not (Test-Path $bindir)) {
-  Write-Error "The output directory `"$bindir`" was not found.`n You must first build the `"SamplePlugin.csproj`" project before calling this script.";
+  Write-Error "The output directory `"$bindir`" was not found.`n You must first build the `"_StreamDeckPlugin_`" project before calling this script.";
```
Comment (NotMyself) [inline scripts/registration/RegisterPluginAndStartStreamDeck.ps1]: Here is another one.

**Comment by csharpfritz** [comment]
Can we just reference *.csproj since there is only one csproj file in the folder?

**Comment by NotMyself** [comment]
I just added a .gitattributes file because I noticed some whitespace munging. It should help.

**Pull request merged**